src/infra/core/src/repos/metadata_chain_impl.rs

```diff
@@ -720,14 +565,18 @@ where
         opts: AppendOpts<'a>,
     ) -> Result<Multihash, AppendError> {
         if opts.validation == AppendValidation::Full {
-            self.validate_append_prev_block_exists(&block).await?;
-            self.validate_append_sequence_numbers_integrity(&block)
-                .await?;
-            self.validate_append_seed_block_order(&block)?;
-            self.validate_append_system_time_is_monotonic(&block)
-                .await?;
-            self.validate_append_watermark_is_monotonic(&block).await?;
-            self.validate_append_offsets_are_sequential(&block).await?;
+            let validators: BoxedVisitors = vec![
+                Box::<ValidateSeedBlockOrderVisitor>::default(),
+                Box::<ValidatePrevBlockExistsVisitor>::default(),
+                Box::<ValidateSequenceNumbersIntegrityVisitor>::default(),
+                Box::<ValidateSystemTimeIsMonotonicVisitor>::default(),
+                Box::<ValidateWatermarkIsMonotonicVisitor>::default(),
+                Box::<ValidateOffsetsAreSequentialVisitor>::default(),
+                // TODO add ValidateEventLogicalStructureVisitor
+            ];
+
+            self.accept(&block, validators).await?;
```

Comment:
We probably need two versions of `accept`:
- one that iterates over existing blocks (public method `accept(visitors)`)
- one that first feeds visitors one extra block that is about to be part of the chain (**private** method `validate(block, visitors)`

Reply:
**Q:** could you please clarify -- can we dispense with the two methods of the Visitor itself? 

(same as https://github.com/kamu-data/kamu-cli/pull/523#discussion_r1506383049)